vulcan_geometry.py

```diff
@@ -0,0 +1,259 @@
+def readPositions(filename: str = CSV_FILE):
+    '''The CSV file has measurements of the front tubes of each 8-pack'''
+    # read in and delete unneccessary columns
+    positions = readFile(filename, delimiter=',')
+    del positions['L']  # tube length
+    del positions['C']  # tube centers
+
+    # set up a dict to split the data into
+    banks_allpixels = {'bank1': {},
+                       'bank2': {},
+                       'bank5': {}}
+    for bank_label in banks_allpixels.keys():
+        for column in positions.keys():
+            banks_allpixels[bank_label][column] = []
+
+    # split the data into banks
+    for i, point_label in enumerate(positions['Point']):
+        bank_label = ''
+        if point_label.startswith('HA_'):
+            bank_label = 'bank5'
+        elif point_label.startswith('BR_'):
+            bank_label = 'bank1'
+        elif point_label.startswith('BL_'):
+            bank_label = 'bank2'
+        if bank_label:
+            banks_allpixels[bank_label]['Point'].append(point_label.split('_')[-1])
+            banks_allpixels[bank_label]['X'].append(positions['X'][i])
+            banks_allpixels[bank_label]['Y'].append(positions['Y'][i])
+            banks_allpixels[bank_label]['Z'].append(positions['Z'][i])
+
+    # convert the positions to ndarray of float
+    for bank_label in banks_allpixels.keys():
+        for column in ['X', 'Y', 'Z']:
+            banks_allpixels[bank_label][column] = np.array(banks_allpixels[bank_label][column], dtype=float)
+
+    def pointFromName(pixels, label: str) -> Vector:
+        index = pixels['Point'].index(label)
+        return Vector(pixels['X'][index], pixels['Y'][index], pixels['Z'][index])
+
+    def rectangleFromName(pixels, corners) -> Rectangle:
+        points = [pointFromName(pixels, name) for name in corners]
+        return Rectangle(*points, tolerance_len=0.035)
+
+    # get the 4 corners into a structure to return
+    # corners in the order LL, UL, UR, LR
+    # Rectangle says lower-left then clockwise
+    # survey/alignment labeled them as "tube 1" (i.e. T1) as most downstream
+    banks = {}
+    banks['bank1'] = rectangleFromName(banks_allpixels['bank1'],
+                                       ('D1T1B', 'D1T1T', 'D20T4T', 'D20T1B'))
+    banks['bank2'] = rectangleFromName(banks_allpixels['bank2'],
+                                       ('D20T1B', 'D20T4T', 'D1T1T', 'D1T1B'))
+    banks['bank5'] = rectangleFromName(banks_allpixels['bank5'],
+                                       ('D1T1B', 'D1T1T', 'D9T4T', 'D9T4B'))
+
+    return banks
+
+
+def addEightPack(instr, name: str, tube_type: str, upsidedown: bool = False):
+    '''Add an interleaved 8-pack where pixel 1 is in the lower-left
+    corner in the back then increases along the tube upward. The bottom
+    of tube two is in the front.
+
+    back     1 3 5 7
+    front     2 4 6 8
+
+    Setting ``upsidedown=True`` configures the pixels as
+
+    back      1 3 5 7
+    front    2 4 6 8
```

Comment:
when I see `upsidedown`, my immediate impression is that the panels would looks like
```
back        2    4    6    8
front    1     3    5    7
```

Reply:
You are correct. The documentation was wrong.